File: lib/posture.ts
"use client"

import { useEffect, useRef, useState, useCallback } from "react"

// Public status shape
export type PostureStatus = {
  score: number
  warnings: string[]
  raw: {
    forwardDiff?: number
    sideDiff?: number
    angleDiff?: number
  }
}

/**
 * Hook: usePostureMonitor
 * Wraps MediaPipe Pose in the browser, reproducing Python logic (baseline diffs + sliding windows).
 */
export function usePostureMonitor(
  videoEl: HTMLVideoElement | null,
  isActive: boolean,
  isPaused: boolean,
  opts: { fps?: number; overlayCanvas?: HTMLCanvasElement | null; resetOnStop?: boolean } = {}
) {
  const poseRef = useRef<any | null>(null)
  const lastLandmarksRef = useRef<any[] | null>(null)
  const baselineRef = useRef<{
    head_forward: number
    head_side_slouch: number
    head_angle: number
  } | null>(null)

  // Sliding window logic mirrors Python's constants
  const WINDOW_SIZE = 100
  const REQUIRED_BAD = 90
  const slouchWindow = useRef<number[]>([])
  const sideWindow = useRef<number[]>([])
  const headWindow = useRef<number[]>([])

  const [status, setStatus] = useState<PostureStatus>({ score: 100, warnings: [], raw: {} })
  const [baselineSet, setBaselineSet] = useState(false)

  const targetFPS = opts.fps ?? 30
  const resetOnStop = opts.resetOnStop ?? true
  const frameInterval = 1000 / targetFPS
  const lastFrameTs = useRef(0)
  const overlayCanvasRef = useRef<HTMLCanvasElement | null>(null)
  overlayCanvasRef.current = opts.overlayCanvas ?? null

  const drawingUtilsRef = useRef<{
    drawConnectors: Function
    drawLandmarks: Function
  } | null>(null)

  const angleBetween = (p1: any, p2: any, p3: any) => {
    const v1 = [p1.x - p2.x, p1.y - p2.y]
    const v2 = [p3.x - p2.x, p3.y - p2.y]
    const dot = v1[0] * v2[0] + v1[1] * v2[1]
    const mag1 = Math.hypot(...v1)
    const mag2 = Math.hypot(...v2)
    if (!mag1 || !mag2) return 0
    const cos = Math.min(1, Math.max(-1, dot / (mag1 * mag2)))
    return (Math.acos(cos) * 180) / Math.PI
  }

  const computeBaseline = useCallback((lm: any[]) => {
    if (!lm) return
    const nose = lm[0]
    const leftEar = lm[7]
    const rightShoulder = lm[12]
    const leftShoulder = lm[11]
    const leftHip = lm[23]
    const head_forward = rightShoulder.x - leftEar.x
    const head_side_slouch = Math.abs(leftEar.z - leftHip.z)
    const head_angle = angleBetween(leftShoulder, leftEar, nose)
    baselineRef.current = { head_forward, head_side_slouch, head_angle }
    setBaselineSet(true)
    // Reset windows & warnings
    slouchWindow.current = []
    sideWindow.current = []
    headWindow.current = []
    setStatus(s => ({ ...s, warnings: [] }))
  }, [])

  const pushWindow = (ref: React.MutableRefObject<number[]>, v: number) => {
    if (ref.current.length >= WINDOW_SIZE) ref.current.shift()
    ref.current.push(v)
  }

  const captureBaseline = useCallback(() => {
    if (lastLandmarksRef.current) computeBaseline(lastLandmarksRef.current)
  }, [computeBaseline])

  const clearWarnings = () => setStatus(s => ({ ...s, warnings: [] }))

  useEffect(() => {
    if (!videoEl || !isActive) return
    let cancelled = false
    let rafId = 0

    const init = async () => {
      try {
  const mp = await import("@mediapipe/pose")
        if (cancelled) return
        if (poseRef.current) return // already initialized (hot reload guard)
        const { Pose } = mp
        // Explicitly pin the version to avoid CDN 'latest' race leading to asset map mismatch
        const POSE_VERSION = '0.5.1675469404'
        const pose = new Pose({
          locateFile: (file: string) => `https://cdn.jsdelivr.net/npm/@mediapipe/pose@${POSE_VERSION}/${file}`
        })
        pose.setOptions({
          modelComplexity: 1,
          enableSegmentation: false,
          minDetectionConfidence: 0.7,
          minTrackingConfidence: 0.7
        })

        // Await internal WASM / assets init to reduce race conditions
        if ((pose as any).initialize) {
          try { await (pose as any).initialize() } catch (e) { console.warn('Pose initialize failed (continuing)', e) }
        }

        // Lazy load drawing utils only if an overlay canvas was provided
        if (overlayCanvasRef.current && !drawingUtilsRef.current) {
          try {
            const du = await import('@mediapipe/drawing_utils')
            drawingUtilsRef.current = {
              drawConnectors: du.drawConnectors,
              drawLandmarks: du.drawLandmarks,
            }
          } catch (e) {
            console.warn('Failed to load drawing utils', e)
          }
        }

        pose.onResults((res: any) => {
          try {
            if (cancelled || isPaused) return
            const lm = res?.poseLandmarks
            if (!lm) return
            lastLandmarksRef.current = lm
            // We still draw landmarks even if baseline not yet set
            const baselineAvailable = !!baselineRef.current

            const nose = lm[0]
            const leftEar = lm[7]
            const rightShoulder = lm[12]
            const leftShoulder = lm[11]
            const leftHip = lm[23]

            const head_forward = rightShoulder.x - leftEar.x
            const head_side_slouch = Math.abs(leftEar.z - leftHip.z)
            const head_angle = angleBetween(leftShoulder, leftEar, nose)

            let forward_diff = 0, side_diff = 0, angle_diff = 0
            let is_slouch = false, is_side = false, is_head = false
            if (baselineAvailable) {
              const base = baselineRef.current!
              forward_diff = Math.abs(head_forward - base.head_forward)
              side_diff = Math.abs(head_side_slouch - base.head_side_slouch)
              angle_diff = Math.abs(head_angle - base.head_angle)

              is_slouch = forward_diff > 0.01
              is_side = side_diff > 0.05
              is_head = angle_diff > 10

              pushWindow(slouchWindow, is_slouch ? 1 : 0)
              pushWindow(sideWindow, is_side ? 1 : 0)
              pushWindow(headWindow, is_head ? 1 : 0)
            }

            const warnings: string[] = []
            if (slouchWindow.current.length === WINDOW_SIZE && slouchWindow.current.reduce((a,b)=>a+b,0) >= REQUIRED_BAD)
              warnings.push('Forward lean detected')
            if (sideWindow.current.length === WINDOW_SIZE && sideWindow.current.reduce((a,b)=>a+b,0) >= REQUIRED_BAD)
              warnings.push('Side lean detected')
            if (headWindow.current.length === WINDOW_SIZE && headWindow.current.reduce((a,b)=>a+b,0) >= REQUIRED_BAD)
              warnings.push('Head lowered detected')

            let score = 100
            if (baselineAvailable) {
              if (is_slouch) score -= 20
              if (is_side) score -= 15
              if (is_head) score -= 15
              score = Math.max(0, score)
            }

            if (baselineAvailable) {
              setStatus(s => ({
                ...s,
                score,
                warnings,
                raw: { forwardDiff: forward_diff, sideDiff: side_diff, angleDiff: angle_diff }
              }))
            }

            // Drawing overlay
            const overlay = overlayCanvasRef.current
            if (overlay && drawingUtilsRef.current) {
              const ctx2 = overlay.getContext('2d')
              if (ctx2) {
                // Ensure canvas matches underlying video size
                if (overlay.width !== videoEl.videoWidth || overlay.height !== videoEl.videoHeight) {
                  overlay.width = videoEl.videoWidth
                  overlay.height = videoEl.videoHeight
                }
                ctx2.save()
                ctx2.clearRect(0,0, overlay.width, overlay.height)
                try {
                  drawingUtilsRef.current.drawConnectors(ctx2, lm, (mp as any).POSE_CONNECTIONS, { color: '#ffffffff', lineWidth: 2 })
                  drawingUtilsRef.current.drawLandmarks(ctx2, lm, { color: '#ff0a0aff', radius: 1 })
                } catch (e) {
                  // Avoid spamming if drawing fails
                }
                ctx2.restore()
              }
            }
          } catch (err) {
            console.error('Pose onResults processing error', err)
          }
        })
        poseRef.current = pose

        const canvas = document.createElement("canvas")
        const ctx = canvas.getContext("2d")

        const startLoop = () => {
          const loop = (ts: number) => {
            if (cancelled) return
            if (isPaused) { rafId = requestAnimationFrame(loop); return }
            if (videoEl.videoWidth > 0 && (ts - lastFrameTs.current >= frameInterval)) {
              lastFrameTs.current = ts
              canvas.width = videoEl.videoWidth
              canvas.height = videoEl.videoHeight
              ctx?.drawImage(videoEl, 0, 0, canvas.width, canvas.height)
              pose.send({ image: canvas })
            }
            rafId = requestAnimationFrame(loop)
          }
          rafId = requestAnimationFrame(loop)
        }
        if (videoEl.readyState >= 2) {
          startLoop()
        } else {
          const onMeta = () => { startLoop(); videoEl.removeEventListener('loadedmetadata', onMeta) }
          videoEl.addEventListener('loadedmetadata', onMeta)
        }
      } catch (e) {
        console.error("Pose init failed", e)
      }
    }

    init()
    return () => {
      cancelled = true
      if (rafId) cancelAnimationFrame(rafId)
      if (poseRef.current?.close) {
        try { poseRef.current.close() } catch {}
      }
      poseRef.current = null
    }
  }, [videoEl, isActive, isPaused, frameInterval])

  // Clear overlay & optionally reset when deactivated
  useEffect(() => {
    if (!isActive) {
      const overlay = overlayCanvasRef.current
      if (overlay) {
        const ctx = overlay.getContext('2d')
        if (ctx) ctx.clearRect(0,0, overlay.width, overlay.height)
      }
      if (resetOnStop) {
        baselineRef.current = null
        setBaselineSet(false)
        setStatus({ score: 100, warnings: [], raw: {} })
        slouchWindow.current = []
        sideWindow.current = []
        headWindow.current = []
      }
    }
  }, [isActive])

  return {
    status,
    baselineSet, // useful for conditionally showing score and badge
    captureBaseline,
    clearWarnings,
  }
}

File: components/ui/posture.py
import cv2
import mediapipe as mp
import math
from collections import deque

mp_drawing = mp.solutions.drawing_utils
mp_pose = mp.solutions.pose

cap = cv2.VideoCapture(0)

def angle_between(p1, p2, p3):
    # angle at p2 formed by p1-p2-p3
    v1 = (p1.x - p2.x, p1.y - p2.y)
    v2 = (p3.x - p2.x, p3.y - p2.y)
    dot = v1[0]*v2[0] + v1[1]*v2[1]
    mag1 = math.hypot(*v1)
    mag2 = math.hypot(*v2)
    return math.degrees(math.acos(dot / (mag1 * mag2)))

# Store baseline posture here after calibration
baseline = None

with mp_pose.Pose(
    min_detection_confidence=0.7,
    min_tracking_confidence=0.7,
    model_complexity=1
) as pose:

    WINDOW_SIZE = 100
    REQUIRED_BAD = 80
    slouch_window = deque(maxlen=WINDOW_SIZE)
    side_window = deque(maxlen=WINDOW_SIZE)
    head_window = deque(maxlen=WINDOW_SIZE)

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        image_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        results = pose.process(image_rgb)
        image = cv2.cvtColor(image_rgb, cv2.COLOR_RGB2BGR)

        if results.pose_landmarks:
            lm = results.pose_landmarks.landmark

            nose = lm[mp_pose.PoseLandmark.NOSE.value]
            ear = lm[mp_pose.PoseLandmark.LEFT_EAR.value]
            right_shoulder = lm[mp_pose.PoseLandmark.RIGHT_SHOULDER.value]
            left_shoulder = lm[mp_pose.PoseLandmark.LEFT_SHOULDER.value]
            left_hip = lm[mp_pose.PoseLandmark.LEFT_HIP.value]

            # Metrics for this frame
            head_forward = right_shoulder.x - ear.x
            head_side_slouch = abs(ear.z - left_hip.z)
            head_angle = angle_between(left_shoulder, ear, nose)

            # If user presses "c" → capture baseline posture
            key = cv2.waitKey(10) & 0xFF
            if key == ord('c'):
                baseline = {
                    "head_forward": head_forward,
                    "head_side_slouch": head_side_slouch,
                    "head_angle": head_angle
                }
                print("✅ Baseline posture captured:", baseline)

            if baseline:
                # Compare current frame to baseline with tolerance
                forward_diff = abs(head_forward - baseline["head_forward"])
                side_diff = abs(head_side_slouch - baseline["head_side_slouch"])
                angle_diff = abs(head_angle - baseline["head_angle"])
                print(f"Diffs - Forward: {forward_diff}, Side: {side_diff}, Angle: {angle_diff}")

                is_slouch = forward_diff > 0.01
                is_side_slouch = side_diff > 0.05
                is_head_lowered = angle_diff > 10

                slouch_window.append(1 if is_slouch else 0)
                side_window.append(1 if is_side_slouch else 0)
                head_window.append(1 if is_head_lowered else 0)

                if len(slouch_window) == WINDOW_SIZE and sum(slouch_window) >= REQUIRED_BAD:
                    cv2.putText(image, "Forward Slouch!", (50, 50),
                                cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 0), 2)
                if len(side_window) == WINDOW_SIZE and sum(side_window) >= REQUIRED_BAD:
                    cv2.putText(image, "Side Slouch!", (50, 90),
                                cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 0), 2)
                if len(head_window) == WINDOW_SIZE and sum(head_window) >= REQUIRED_BAD:
                    cv2.putText(image, "Head Lowered!", (50, 130),
                                cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 0), 2)
                if len(slouch_window):
                    print(f"Slouch window: {sum(slouch_window)}/{len(slouch_window)}")
                if len(side_window):
                    print(f"Side window: {sum(side_window)}/{len(side_window)}")
                if len(head_window):
                    print(f"Head window: {sum(head_window)}/{len(head_window)}")

            mp_drawing.draw_landmarks(image, results.pose_landmarks, mp_pose.POSE_CONNECTIONS)

        cv2.imshow("Posture Monitor", image)
        if cv2.waitKey(10) & 0xFF == ord('q'):
            break

cap.release()
cv2.destroyAllWindows()
